File: 백준파이썬/중고급1/다이나믹프로그래밍5/우수마을.py
import sys
sys.setrecursionlimit(10**6)
sys.stdin = open('1949.txt', 'r')
input = sys.stdin.readline

N = int(input())
W = [0] + list(map(int, input().split()))
tree = [[] for i in range(N+1)]
for i in range(N-1):
    a, b = map(int, input().split())
    tree[a].append(b)
    tree[b].append(a)
DP = [[0, 0] for i in range(N+1)]


def dfs(u, parent):
    DP[u][0] = 0
    DP[u][1] = W[u]
    for v in tree[u]:
        if v != parent:
            dfs(v, u)
            DP[u][0] += max(DP[v][0], DP[v][1])
            DP[u][1] += DP[v][0]


dfs(1, 0)
print(max(DP[1][0], DP[1][1]))

File: 2021/03/0331js/우수마을.js
// const input = require('fs').readFileSync('/dev/stdin').toString().split("\n");
// const input = require('fs').readFileSync('dev/stdin').toString().split('\n').map(x=>x.split(' ').map(x=>x.trim()));

const input = require("fs").readFileSync("./1949.txt").toString().split("\n");
const N = parseInt(input[0]);
const cost = [0, ...input[1].split(" ").map(Number)];
const dp = Array.from(Array(N + 1), () => Array(2).fill(0));
const tree = Array.from(Array(N + 1), () => new Array());
for (let i = 2; i < 2 + N - 1; i++) {
  const [a, b] = input[i].split(" ").map(Number);
  tree[a].push(b);
  tree[b].push(a);
}

const dfs = (prev, u) => {
  dp[u][0] = 0;
  dp[u][1] = cost[u];
  for (let v of tree[u]) {
    if (prev !== v) {
      dfs(u, v);
      dp[u][0] += Math.max(dp[v][0], dp[v][1]);
      dp[u][1] += dp[v][0];
    }
  }
};

dfs(0, 1);
console.log(Math.max(dp[1][0], dp[1][1]));
